### ISSUE TITLE
Data set generators 

### ISSUE THREAD

**Issue by mloning** [alan-turing-institute/sktime#353]
As discussed in #320, it would be great to have functions that generate/simulate data sets, they are useful for testing and explaining the difference between estimators. 

* We currently have a very simple data set generator here: [`make_classification_problem`](https://github.com/alan-turing-institute/sktime/blob/67c56be8b1e838f2628df829946f795b7dba9aed/sktime/utils/_testing/series_as_features.py#L33)
* TSML has some here in Java: https://github.com/uea-machine-learning/tsml/tree/master/src/main/java/statistics/simulators

**Comment by Kerybas** [comment]
I am interested in this issue.  
I would like to propose the following approach:  
1. Generate a bunch of randomised sine waves (10 here):
![image](https://user-images.githubusercontent.com/42703200/89106120-92cddf00-d427-11ea-8c07-e54d6df290c4.png)
2. Pass them through a sklearn MLPRegressor randomly fitted, to obtain a number of "observations":
![image](https://user-images.githubusercontent.com/42703200/89106156-d1fc3000-d427-11ea-9a19-760ce8e7413f.png)
3. Optionally, discretize some of the signals in 2. to obtains classes (here, the red signal is discretized in 3 bins via KBinsDiscretizer):
![image](https://user-images.githubusercontent.com/42703200/89106248-8d24c900-d428-11ea-99d5-5a442ed4cce0.png)  

Step 2. should ensure that the data set is complex while keeping a realistic relationship between the signals.  

We could also add some noise, under-sampling, etc. to step 2 to make it even more realistic.

What do you think?  
Is it worth investigating that approach?

**Comment by mloning** [comment]
@ABostrom @TonyBagnall any thoughts on this?

**Comment by ABostrom** [comment]
There's an arxiv paper did ages ago here: https://arxiv.org/abs/1703.09480
 
An old SimulationExperiment class from tsml: https://github.com/uea-machine-learning/tsml/blob/master/src/main/java/experiments/SimulationExperiments.java

I think our angle for simulation was, what type of problem are you looking to simulate within TSC.

Can you design a simulated problem that an interval classifier should be best on, same for DTW, dictionary, shapelet etc.

**Comment by fkiraly** [comment]
Hm, @TonyBagnall, @Kerybas, I thought it would be useful to get a quick way to generate synthetic test datasets for the different *tasks*, e.g., time series classification, regression, forecasting, that sktime supports?

**Comment by mloning** [comment]
Simulators for discriminating between classifiers would be useful for the part of the user guide #361 where we want to guide algorithm choice and explain the inner workings of the different algorithms.

**Comment by TonyBagnall** [comment]
so I wrote the data simulator to reinforce the design mechanism of hive-cote,  i.e. that if you have no prior knowledge as to the correct feature space, a meta ensemble is the best approach. A simulator that just creates a task problem is much simpler, simply generate white noise. However, I would like to use more simulation experimentation, not least to satisfy the statisticians. I think we could definitely add this to the new road map we are developing, and make it a project, once we have prototyped the code structure

**Comment by mloning** [comment]
Gluon-ts also have a lot of simulation functionality for forecasting/anomaly detection: https://gluon-ts.mxnet.io/examples/synthetic_data_generation_tutorial/tutorial.html

**Comment by sjmiller8182** [comment]
I have existing code for simulating univariate ARMA, ARIMA time series based on a given polynomial. Would that be of interest here?

**Comment by mloning** [comment]
@sjmiller8182 yes certainly, the discussion so far has focussed on time series classification, but having data generators for forecasting would be very helpful too! 👍 

I'd appreciate a draft PR with a new file for generators in [`sktime/datasets/`](https://github.com/alan-turing-institute/sktime/tree/master/sktime/datasets). Also feel free to open a separate issue to discuss this in more detail if you have more questions! 

FYI statsmodels has an [`ARMAProcess`](https://www.statsmodels.org/stable/generated/statsmodels.tsa.arima_process.ArmaProcess.html#statsmodels.tsa.arima_process.ArmaProcess) generator too.

**Comment by karthikn72** [comment]
Hey, could I please know about the progress with this? If this is still ongoing, could I perhaps have a go at this? I am new to making open-source contributions and would love to get started this way :)

**Comment by mloning** [comment]
Hi @karthikn72 - yes this is still open, would be a great issue to work on as a starting point I think.

**Comment by TonyBagnall** [comment]
hi @karthikn72 yes it would be good, if you want to do it for classification problems I'm happy to help.

**Comment by karthikn72** [comment]
@TonyBagnall Sure, this seems like a nice problem for me to work on. I'll have a deeper look at the links above over the weekend. Would you say it's a good idea to look around for current implementations of data simulators for models that can be classified by sktime, for e.g. shapelet-based or interval-based models?

**Comment by TonyBagnall** [comment]
Hi, I think starting with a simulator for one type of representation would be good yes, and trying to modularise it so it's easy to do others is worthwhile.. look at tsml but don't feel you have to follow the same structure, I'll discuss that with the team on monday

**Comment by karthikn72** [comment]
Alright, I had a look at tsml and will start by reimplementing the Shapelet simulator for now to understand what it does, and see where I can go from there. I'll aim to open a PR ASAP so we can continue discussing there :)
